subsys/bluetooth/audio/csip_set_coordinator.c

```diff
@@ -1568,10 +1549,40 @@ static int verify_members(const struct bt_csip_set_coordinator_set_member **memb
 	return 0;
 }
 
-static int bt_csip_set_coordinator_get_lock_state(
-	const struct bt_csip_set_coordinator_set_member **members,
-	uint8_t count,
-	const struct bt_csip_set_coordinator_set_info *set_info)
+static bool check_and_set_members_busy(const struct bt_csip_set_coordinator_set_member *members[],
+				       size_t count)
+{
+	size_t num_free;
+
+	for (num_free = 0U; num_free < count; num_free++) {
+		const struct bt_csip_set_coordinator_set_member *member = members[num_free];
+		struct bt_csip_set_coordinator_inst *client =
+			CONTAINER_OF(member, struct bt_csip_set_coordinator_inst, set_member);
+
+		if (atomic_test_and_set_bit(client->flags, SET_COORDINATOR_FLAG_BUSY)) {
+			LOG_DBG("Member[%zu] (%p) is busy", num_free, member);
+			break;
+		}
+	}
+
+	/* If any is busy, revert any busy states we've set */
+	if (num_free != count) {
+		for (size_t i = 0U; i < num_free; i++) {
+			const struct bt_csip_set_coordinator_set_member *member = members[i];
+			struct bt_csip_set_coordinator_inst *client = CONTAINER_OF(
+				member, struct bt_csip_set_coordinator_inst, set_member);
+
+			atomic_clear_bit(client->flags, SET_COORDINATOR_FLAG_BUSY);
+		}
+	}
```

Comment:
Could all of this logic perhaps be simplified by having a per-coordinator set bitmap comprising of the busy states of all its members? That way you can test with O(1) complexity that the entire bitmap is zero or that at least one member is busy, e.g. `if (atomic_get(&busy) == 0)`.

That said, there's a lot of code here that looks problematic from a multithreading perspective. The implication of using `test_and_set_bit()` is that you can be preempted in this context and other threads may modify e.g. `client->flags`. If that's the case, then this simply isn't thread safe. You'd need to have a per-client lock to ensure that the full critical region is protected.

Reply:
I might be missing something here. 

> Could all of this logic perhaps be simplified by having a per-coordinator set bitmap comprising of the busy states of all its members?

We want to clock individual clients/set coordinators, so the lock should exist only for that. The resources that we want to lock is per client/coordinator, and not per member. 

> That way you can test with O(1) complexity that the entire bitmap is zero or that at least one member is busy, e.g. if (atomic_get(&busy) == 0).

From just looking at this function, I see how it's a bit confusing. In short each client/coordinator may have multiple members. The procedures in CSIP work on one or more members from different clients. Before we call this function, we always verify that with the `verify_members` functions, that ensure that no client has more than 1 member in the parameters. 
So what we check and set here, is that each each member (that do not share clients/coordinators) has a client that is not busy. For that, we always need to verify all the clients of the members, which cannot be done in O(1). Even if we did have a global busy flag state, the clients in a specific procedure may only be subset of all clients, and as such to generate the bitmask, we still need to iterate on `count`. 
The get/set could be done in a single action, so there is some improvement in terms of processing and memory, by using a global `atomic` state variable for all clients. Let me think about that as we can just use the `bt_conn.index` for the bits. 

> That said, there's a lot of code here that looks problematic from a multithreading perspective. The implication of using test_and_set_bit() is that you can be preempted in this context and other threads may modify e.g. client->flags. If that's the case, then this simply isn't thread safe. You'd need to have a per-client lock to ensure that the full critical region is protected.

Can you elaborate on this? The lock is per client in the current version

Reply:
There are indeed some memory optimization that can done, especially if we just need a busy flag. For example if we have 8 connections, we can manage that with a single `uint8_t` rather than one per client. A small but not insignificant optimization. 

The biggest concern I have about a global state is that we are moving ownership of the state out from the struct, and into a global one. It can certainly be managed, but it does seem a bit weird to do, and will weirder for each additional flag (if any) we add. 

It is also possible to optimize the loops here as we can set and restore multiple bits in fewer `atomic` calls, but requires the above global state variable instead of a struct specific one. I think if we go with that approach for any of the LE Audio implementations, we should apply it to all of them (which again is an optimization, but at a small cost in terms of readability and proper ownership IMO). 

Please let me know your thoughts (and other reviewers as well of course)

Reply:
> > That said, there's a lot of code here that looks problematic from a multithreading perspective. The implication of using test_and_set_bit() is that you can be preempted in this context and other threads may modify e.g. client->flags. If that's the case, then this simply isn't thread safe. You'd need to have a per-client lock to ensure that the full critical region is protected.
> 
> Can you elaborate on this? The lock is per client in the current version

It seemed initially that the two for-loops depend on not being preempted anywhere in between (e.g. by another thread entering the same function), however now that I looked at it more it seems like this should indeed be fine, i.e. the second loop will clear only bits set by the first loop in the same invocation of the function.